File: DSA/PranyaGupta_500126032/Day26Question1.py
class node:
    def __init__(self, data):
        self.next = None
        self.data = data
class LinkedList:
    def __init__(self):
        self.head = None
    def add(self, data):
        new_node = node(data)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node
    def length(self,head):
        l = 0
        curr = head
        while curr:
            l+=1
            curr= curr.next
        return l 
    def printlist(self, head):
        current = head
        while current:
            print(current.data, end=" ")
            current = current.next     
         
    def swapkth(self, head, k):
        l = self.length(head)
        if k>l or 2*k-1 == l:
            return head
        if k == l-k:
            return head
        lcurr = head
        lprev = None
        rcurr = head
        rprev = None
        for i in range(k-1):
            lprev = lcurr
            lcurr = lcurr.next
        for i in range(l-k):
            rprev = rcurr
            rcurr = rcurr.next
        if lprev:  
            lprev.next = rcurr
        else:
            head = rcurr
        
        if rprev:
            rprev.next = lcurr
        else:
            head = lcurr    
        temp = lcurr.next
        lcurr.next = rcurr.next
        rcurr.next = temp
        
        self.printlist(head)
def main():
    user_input = input("Enter elements of the linked list (with space in between): ")
    arr = list(map(int, user_input.strip().split()))
    k = int(input("Enter k: "))
    l = LinkedList()
    for n in arr:
        l.add(n)
    l.swapkth(l.head, k)
    
if __name__ == "__main__":
    main()

File: DSA/GEHNA_KHANNA_590013140/day26question1.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* create(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

struct Node* swapKth(struct Node* head, int k) {
    int n = 0;
    struct Node* temp = head;
    while (temp) {
        n++;
        temp = temp->next;
    }
    if (k > n || 2 * k - 1 == n) return head;

    struct Node *x = head, *x_prev = NULL;
    for (int i = 1; i < k; i++) {
        x_prev = x;
        x = x->next;
    }

    struct Node *y = head, *y_prev = NULL;
    for (int i = 1; i < n - k + 1; i++) {
        y_prev = y;
        y = y->next;
    }

    if (x_prev) x_prev->next = y;
    if (y_prev) y_prev->next = x;

    struct Node* temp2 = x->next;
    x->next = y->next;
    y->next = temp2;

    if (k == 1) head = y;
    if (k == n) head = x;

    return head;
}

void printList(struct Node* head) {
    while (head) {
        printf("%d", head->data);
        if (head->next) printf("->");
        head = head->next;
    }
    printf("\n");
}

int main() {
    int n, k, val;
    struct Node *head = NULL, *curr = NULL;

    printf("Enter number of nodes: ");
    scanf("%d", &n);

    printf("Enter elements: ");
    for (int i = 0; i < n; i++) {
        scanf("%d", &val);
        struct Node* newNode = create(val);
        if (head == NULL) {
            head = newNode;
            curr = head;
        } else {
            curr->next = newNode;
            curr = curr->next;
        }
    }

    printf("Enter k: ");
    scanf("%d", &k);

    head = swapKth(head, k);
    printList(head);

    return 0;
}
